timm/models/efficientvit_msra.py

```diff
@@ -0,0 +1,520 @@
+                                 window_resolution=window_resolution, kernels=kernels))
+
+        self.dw1 = ResidualDrop(ConvBN(ed, ed, 3, 1, 1, groups=ed, bn_weight_init=0.))
+        self.ffn1 = ResidualDrop(FFN(ed, int(ed * 2)))
+
+    def forward(self, x):
+        return self.ffn1(self.dw1(self.mixer(self.ffn0(self.dw0(x)))))
+
+
+class PatchEmbedding(torch.nn.Sequential):
+    def __init__(self, in_chans, dim):
+        super().__init__()
+        self.add_module('conv1', ConvBN(in_chans, dim // 8, 3, 2, 1))
+        self.add_module('relu1', torch.nn.ReLU())
+        self.add_module('conv2', ConvBN(dim // 8, dim // 4, 3, 2, 1))
+        self.add_module('relu2', torch.nn.ReLU())
+        self.add_module('conv3', ConvBN(dim // 4, dim // 2, 3, 2, 1))
+        self.add_module('relu3', torch.nn.ReLU())
+        self.add_module('conv4', ConvBN(dim // 2, dim, 3, 2, 1))
+        self.patch_size = 16
+
+
+class EfficientViTMSRA(nn.Module):
+    def __init__(
+        self,
+        img_size=224,
+        in_chans=3,
+        num_classes=1000,
+        embed_dim=[64, 128, 192],
+        key_dim=[16, 16, 16],
+        depth=[1, 2, 3],
+        num_heads=[4, 4, 4],
+        window_size=[7, 7, 7],
+        kernels=[5, 5, 5, 5],
+        down_ops=[[''], ['subsample', 2], ['subsample', 2]],
+        global_pool='avg',
+    ):
+        super(EfficientViTMSRA, self).__init__()
+        self.grad_checkpointing = False
+        resolution = img_size
+        # Patch embedding
+        self.patch_embed = PatchEmbedding(in_chans, embed_dim[0])
+        stride = self.patch_embed.patch_size
+        resolution = img_size // self.patch_embed.patch_size
+
+        attn_ratio = [embed_dim[i] / (key_dim[i] * num_heads[i]) for i in range(len(embed_dim))]
+        self.feature_info = []
+        stages = []
+
+        # Build EfficientViT blocks
+        for i, (ed, kd, dpth, nh, ar, wd, do) in enumerate(
+                zip(embed_dim, key_dim, depth, num_heads, attn_ratio, window_size, down_ops)):
+            blocks = []
+            if do[0] == 'subsample' and i != 0:
+                # Build EfficientViT downsample block
+                resolution_ = (resolution - 1) // do[1] + 1
+                blocks.append(torch.nn.Sequential(ResidualDrop(ConvBN(embed_dim[i - 1], embed_dim[i - 1], 3, 1, 1, groups=embed_dim[i - 1])),
+                              ResidualDrop(FFN(embed_dim[i - 1], int(embed_dim[i - 1] * 2))),))
+                blocks.append(PatchMerging(*embed_dim[i - 1:i + 1]))
+                resolution = resolution_
+                blocks.append(torch.nn.Sequential(ResidualDrop(ConvBN(embed_dim[i], embed_dim[i], 3, 1, 1, groups=embed_dim[i])),
+                              ResidualDrop(FFN(embed_dim[i], int(embed_dim[i] * 2))),))
+                stride *= 2
+            for d in range(dpth):
+                blocks.append(EfficientViTBlock(ed, kd, nh, ar, resolution, wd, kernels))
+            stages.append(nn.Sequential(*blocks))
```

Comment:
As with my Levit refactor, the downsample (patch merging in this model) should really be pushed into a stage module along with the corresponding blocks for the stage https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/levit.py#L461 ... it makes it easier to work with the network

Reply:
done